tests/integration/upsertGraph.js

```diff
@@ -1777,6 +1778,110 @@ module.exports = session => {
           });
       });
     });
+    describe('upserts with update: true option', () => {
+      before(() => {
+        Model1.$$jsonSchema = {
+          type: 'object',
+          required: ['model1Prop1'],
+
+          properties: {
+            model1Prop1: {type: 'string'}
+          }
+        };
+      });
+
+      after(() => {
+        delete Model1.$$jsonSchema;
+        delete Model1.$$validator;
+      });
+
+      it('should fail to do an incomplete upsert', () => {
+        const fails = [
+          {
+            id: 2,
+            // This fails because of missing property.
+            // model1Prop1: 'updated root 2',
```

Comment:
There's an optimization that no updates are done if there is only an identifier present. Your code never reaches the validation line. A better way to test this would be to set the `model1Prop1` and give it a value of wrong type.

Reply:
Oh I see, that makes sense. But I need to catch the case where an update does not provide all required properties and hence causes that kind of validation error, which a patch wouldn't cause due to the use of `skipRequired` in compileValidator() (which as far as I can tell is the only difference between patch and update?)

Reply:
The optimization is in `UpsertNode.hasChanges` it tests if any of the values has actually changed. Since we only fetch ids and foreign keys from the db, the test only compares those. I think the way it currently works is correct. If there is nothing to update, even if the model is invalid, nothing should be done.

Reply:
We can't move the validation to an earlier position, since some properties get set by the insert part. The validation must be done after insert, but before update.

Reply:
Not saying it's not correct. I guess I'd need a model with more than one property to create an incomplete update?